### PR TITLE
dynamic_one_shot is compatible with jax.jit 

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
As a first step toward our goal to support `dynamic_one_shot` with Catalyst, we attempt to support `jax.jit`.

**Description of the Change:**
Make all array shapes static. Perform post-selection as a filtering operation during the post-processing step of `dynamic_one_shot`. Add `jax.jit` tests. We also introduce the constant `fillin_value` which is used as a placeholder for postselection mismatched samples.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-61173]
[sc-61315]
[sc-62094]

### CODE REVIEW

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:138]: Could you add a docstring to this test?

**Review by mudit2812 — COMMENTED** [review]
Looks great! There are a few tests in test_default_qubit_native_mcm.py that are missing docstrings. Could you add those as well?

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:310]: You're right. We achieve a minimal amount of splitting with some global view of the code, but I wonder whether it wouldn't be better to take a bottom-up view for programmers with less knowledge about the code. This way it does not matter where the key comes from (say a new function by an external contributor), it remains reusable one level up in the stack. That said, I'm curious what you think but I don't really mind either way.

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:313]: Yes indeed.

Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py]: As for `choice` versus `binomial`, we simply seem to use `choice` elsewhere in the code but I'm fine switching to binomial.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (vincentmr) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:138]: Will do.

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py]: Thanks, much cleaner.

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [reply pennylane/devices/qubit/apply_operation.py]: I like your idea, we would also avoid meaningless splits this way as currently the key is split for every op in the tape (though it has no functional implications).

Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py]: You still need to split at the top level because we are not returning the first key. Otherwise we will always obtain the same key here. I'm rather thinking of future code that might call `apply_operation` for some reason and then sample a prob-dist. Then even if the developer doesn't split prior to sampling, the key is still unused. So it is safer to split at both levels albeit sometimes useless. If we want to keep splitting to a minimum tough, I would remove the splitting in `apply_operation` and maybe split in `get_final_state` depending on `isinstance`.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:658]: I don't think this is needed if you mark a test with the `jax` marker.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:689]: 👍

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:335]: Could you add a dev note about why we're using this matrix rather than conditionally applying a `PauliX`?

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:321]: Could you add a note here about why we're using a projector matrix rather than directly applying `qml.Projector`?

```suggestion
```

**Review by mudit2812 — COMMENTED** [review]
Looks great! I'm pretty much ready to approve, but just curious about what we should do about `qml.probs`. I would like us to finalize our resolution regarding `qml.probs` before I approve.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (vincentmr) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:658]: Correct, I'll remove it.

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:321]: Will do.

**Comment by vincentmr** [comment]
Thanks @mudit2812 , I've added a couple notes why we are using `apply_operation(qml.QubitUnitary,...)`. We could dig more to maybe enable `qml.Projector` at some point, but `reset` would require introducing `jax.lax.cond` and I think it's better to keep this to a minimum and pay a small performance penalty. As for `probs`, this should be resolved by #5619, but I thought we would want to make a separate PR since it may have further reaching implications. What do you think?

**Review by mudit2812 — APPROVED** [review]
That sounds good to me! Approving now :)

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:319]:
```suggestion
    # to select the sample branch enables jax.jit and prevents it from using Python callbacks
```

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:330]:
```suggestion
    # to reset enables jax.jit and prevents it from using Python callbacks
```

File: pennylane/transforms/dynamic_one_shot.py
Comment (AmintorDusko) [inline pennylane/transforms/dynamic_one_shot.py:39]:
```suggestion
fill_in_value = np.iinfo(np.int32).min
```
If I understood it right.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (AmintorDusko) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:698]: Do we want to remove it definitely? Or maybe comment why this is commented?

Comment (AmintorDusko) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:698]: Same below.

**Review by AmintorDusko — APPROVED** [review]
I left some minor comments, but it looks pretty good!

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (vincentmr) [reply tests/devices/default_qubit/test_default_qubit_native_mcm.py:698]: Will be fixed in a follow-up PR.

**Pull request merged**
